### PR TITLE
Feature/575/display max indicator in the metric dropdown

### LINKED ISSUE #575: Add search to the metric dropdown and display some form of max Indicator for individual metrics
# Feature request

In the metrics bar CodeCharta currently shows the max value of the metric in brackets `rloc (3000)`. It's not clear that it is max though. 

* Please check if we can add a max Symbol in front of the value or something else to make it clear. 
* Please add search functionality for the metrics as well

### PR DESCRIPTION
**Pull request by JaiJoPa**
# Added indicator for highest value in map for metric

closes #575

## Description

Added ,,highest value'' label before each metric max value in metric chooser. That should indicate, that the highest value, that can be found under the chosen metric, will be that value.

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

**Comment by alschmut** [comment]
That's a bit much information, if we gonna show it on every metric option. How about adding disabled option on top which says "metric (max value)". Is a bit smaller in font size and grey?

**Comment by JaiJoPa** [comment]
I made Screenshots for the two opions. 


Thats how it looks, if we show it on each metric option: 

Metric_Info_each_metric


And thats how it looks if we have a disabled opion on top of the list

metric_info_top_option

**Comment by alschmut** [comment]
I actually like the second screenshot 👍

**Comment by alexhunziker** [comment]
I like the 2nd option too, but i would actually give the "metric (highest val)" field the same height as the other entries

**Comment by alschmut** [comment]
This header also looks nice and comes with a build in search https://material.angularjs.org/latest/demo/select#select-header

**Comment by JaiJoPa** [comment]
@alexhunziker I've got a screenshot to that too

Metric Info Top

@alschmut: yeah I liked that build in search header too.

**Comment by Richargh** [comment]
So, should we use the other header, add search functionality as well and change the issue goal?

**Comment by JaiJoPa** [comment]
Personally I would like the search functionality :)

**Comment by Richargh** [comment]
Me too :) It just wasn't clear what was implemented in this PR which you closed 2h ago. Did you close this PR because we're going in a new direction?

**Comment by JaiJoPa** [comment]
No, I closed it accidently when I commented and didn't know how to reopen it 😅😅😅

**Comment by Richargh** [comment]
I changed the original issue to ** Add search to the metric dropdown and display some form of max Indicator for individual metrics** in #575

**Comment by Richargh** [comment]
also I reoponed the PR :)

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.component.ts [hunk 1]
```diff
@@ -44,8 +47,20 @@ export class MetricChooserController implements MetricServiceSubscriber, CodeMap
 		this.updateViewModel(settings)
 	}
 
+	public filterMetricData() {
+		this._viewModel.metricData = this.metricData.filter(metric =>
+			metric.name.toLowerCase().includes(this._viewModel.searchTerm.toLocaleLowerCase())
```
Comment (NearW) [inline visualization/app/codeCharta/ui/metricChooser/metricChooser.component.ts]: use toLowerCase for the searchTerm as well

**Comment by NearW** [comment]
Add a border to the dropdown so it stands out more when the settings-panel are opened

Screen Shot 2019-08-02 at 11 45 08

**Comment by NearW** [comment]
Remove the empty white div when no metric is found with the search term

Screen Shot 2019-08-02 at 11 45 30

**Comment by alschmut** [comment]
In Safari on a Mac it looks like this. I agree with @NearW's comments
- I would also rather remove the grey border for the search. 
- Maybe add some very light grey background for the search. 
- Did you check, if a search-symbol looks good on the left inside the search-field?

**Comment by alexhunziker** [comment]
I noticed the following behaviour (Windows, Firefox):

We have a metric that is fully visible (undocumented_api, including max count)
grafik


Now when I start searching, and the field gets smaller, cutting of the metrics max count
grafik

**Comment by JaiJoPa** [comment]
I've put a border around the metrics. And made the searchbar gray.  I've made 2 screenshots one with a searchbar border and one without.

The one without searchbar border :

bar_without_border

The one with searchbar border:

bar_with_border

What you think ? 

@alschmut: Didn't checked the idea with the icon yet but I like that Idea.

**Comment by alschmut** [comment]
@JaiJoPa I think I would prefer a shadow around the box without a border. So just add the  `dark_shadow_opacity` CSS class to the box. Its defined in `codecharta-component.scss`

**Comment by JaiJoPa** [comment]
@alschmut Do you mean like this ?:

bar_without_border_2

**Comment by alschmut** [comment]
@JaiJoPa that looks really nice! When focusing the input-field, a dark green border is added below, which moves the placeholder a bit (approx. 2px). Maybe you can add a border by default which has the same color as the grey background. So the green border overwrites it.

- And you need to add the search symbol to the other 3 metric-chooser as well.

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss [hunk 2]
```diff
@@ -74,3 +71,49 @@ distribution-metric-chooser-component {
...
+	}
+
+	&.distributionBarDropdown {
+		top: 5.5% !important;
+	}
+}
+
+md-content.option-group {
+	top: $optionHeight + px !important;
+	min-height: auto;
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss]: Yes we need that to remove the empty white div, if the metric chooser doesn't find a match, which @NearW mentioned

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss [hunk 3]
```diff
@@ -74,3 +71,49 @@ distribution-metric-chooser-component {
...
 	}
 }
+
+.md-select-menu-container {
+	&.ribbonBarDropdown {
+		top: 10% !important;
+	}
+
+	&.distributionBarDropdown {
+		top: 5.5% !important;
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss]: done

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss [hunk 4]
```diff
@@ -74,3 +71,49 @@ distribution-metric-chooser-component {
 		}
 	}
 }
+
+.md-select-menu-container {
+	&.ribbonBarDropdown {
+		top: 10% !important;
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/metricChooser/metricChooser.component.scss]: done

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.area.component.html [hunk 5]
```diff
@@ -1,11 +1,28 @@
...
+	<label></label>
+	<md-select
+		ng-model="$ctrl._viewModel.areaMetric"
+		ng-change="$ctrl.applySettingsAreaMetric()"
+		md-on-close="$ctrl.clearSearchTerm()"
+		md-container-class="ribbonBarDropdown"
+	>
+		<md-select-header>
+			<input
+				class="user-search"
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/metricChooser/metricChooser.area.component.html]: You are right! I changed it to metric-search, which probably describes the input better.

File: visualization/app/codeCharta/ui/metricChooser/metricChooser.component.ts [hunk 6]
```diff
@@ -18,19 +18,22 @@ export class MetricChooserController implements MetricServiceSubscriber, CodeMap
 	public hoveredAreaDelta: number
 	public hoveredColorDelta: number
 	public hoveredDeltaColor: string
+	private metricData: MetricData[]
 
 	private _viewModel: {
 		metricData: MetricData[]
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/metricChooser/metricChooser.component.ts]: Done.

**Pull request merged**